server/sonar-server/src/main/java/org/sonar/server/qualityprofile/RuleActivator.java

```diff
@@ -423,118 +363,119 @@ private String validateParam(RuleParamDto ruleParam, @Nullable String value) {
     return value;
   }
 
-  public BulkChangeResult bulkActivateAndCommit(DbSession dbSession, RuleQuery ruleQuery, QProfileDto profile, @Nullable String severity) {
-    BulkChangeResult result = new BulkChangeResult();
-    Iterator<RuleKey> rules = ruleIndex.searchAll(ruleQuery);
-    while (rules.hasNext()) {
-      RuleKey ruleKey = rules.next();
-      try {
-        RuleActivation activation = RuleActivation.create(ruleKey, severity, null);
-        List<ActiveRuleChange> changes = activate(dbSession, activation, profile);
-        result.addChanges(changes);
-        if (!changes.isEmpty()) {
-          result.incrementSucceeded();
-        }
+  public RuleActivationContext createContextForBuiltInProfile(DbSession dbSession, RulesProfileDto builtInProfile, Collection<RuleKey> ruleKeys) {
+    checkArgument(builtInProfile.isBuiltIn(), "Rules profile with UUID %s is not built-in", builtInProfile.getKee());
 
-      } catch (BadRequestException e) {
-        // other exceptions stop the bulk activation
-        result.incrementFailed();
-        result.getErrors().addAll(e.errors());
-      }
+    RuleActivationContext.Builder builder = new RuleActivationContext.Builder();
+
+    // load rules
+    List<RuleDefinitionDto> rules = completeWithRules(dbSession, builder, ruleKeys);
+
+    // load profiles
+    List<QProfileDto> aliasedBuiltInProfiles = db.qualityProfileDao().selectQProfilesByRuleProfile(dbSession, builtInProfile);
+    List<QProfileDto> profiles = new ArrayList<>(aliasedBuiltInProfiles);
+    for (QProfileDto aliasProfile : aliasedBuiltInProfiles) {
+      // FIXME N+1 syndrome. To be optimized by replacing db column parent_kee by ancestor_keys.
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Take the required action to fix the issue indicated by this comment. ([squid:S1134](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsb%2FSONAR-10052%2Fqprofiles&issues=AWE-qvP5mckxG3vIth77&open=AWE-qvP5mckxG3vIth77)

[ISSUE_KEY]: AWE-qvP5mckxG3vIth77